### DISCUSSION TITLE
Supporting json column type in database

### DISCUSSION THREAD

**Discussion by thachnb85** [simolus3/drift#1087] [General]
Hi @simolus3,
Sqlite has a json extension to store data in json column, https://www.sqlite.org/json1.html
So for following table:
```
create table a (id int primary key, data json)
```
`id` can be represented as IntColumn, how about `data`?
Thank you,

**Comment by thachnb85** [comment]
I tried another way, by creating tables.moor file, and run the generator
I added the data field which has json type
```
-- this is the tables.moor file
CREATE TABLE todos (
    id INT NOT NULL PRIMARY KEY AUTOINCREMENT,
    title TEXT,
    body TEXT,
    data JSON,
    category INT REFERENCES categories (id)
);
```

when I run `flutter packages pub run build_runner build`
I saw some errors:
```
[INFO] 2.6s elapsed, 1/2 actions completed.
[WARNING] moor_generator:moor_generator on lib/services/data_manager/database/database.dart:
There were some errors while running moor_generator on package:test/services/data_manager/database/database.dart:
[WARNING] moor_generator:moor_generator on lib/services/data_manager/database/database.dart:
line 6, column 14: Expected a JSON KEY constraint
  ╷
6 │     data JSON,
  │              ^
  ╵
```

**Reply by simolus3** [reply]
In sqlite3, json isn't an extra column type. You can just store them as `TEXT` which should work. Using `JSON` in a table definition is legal in sqlite3, but that column would have a numeric type (due to [type affinity](https://www.sqlite.org/datatype3.html#determination_of_column_affinity)). In moor, `JSON` is a special keyword to control the `toJson` method of a data class. That may not have been a great idea either, but that's what it is :D

In short, just store json columns as `TEXT` and then use the json functions on that.

**Reply by thachnb85** [reply]
Hi @simolus3, thank you for your reply.
I tested with the example app, by adding one field calling jsonData to TodoEntry
```
@DataClassName('TodoEntry')
class Todos extends Table {
  IntColumn get id => integer().autoIncrement()();
  TextColumn get content => text()();
  DateTimeColumn get targetDate => dateTime().nullable()();
  TextColumn get jsonData => text()();
  IntColumn get category => integer()
      .nullable()
      .customConstraint('NULLABLE REFERENCES categories(id)')();
}
```
Then I tried to insert some test data, all good at this point:
```
          for (int i = 0; i < 10; i ++){
            print('inserting ---------- $i');
            await into(todos).insert(TodosCompanion(
              content: Value("todo entry $i"),
              targetDate: Value(DateTime.now()),
              jsonData: Value('{"index":"$i"}'),
            ));
          }
```

But, when I ran the json query, it doesn't work. For example writing a custom select with json_extract method:

```
  Future<TodoEntry> selectSomething() {
    // This custom select works perfect:
    // return (select(todos)..where((tbl) => tbl.id.equals(2))).getSingle();

    // but this json function doesn't work
    return customSelect(
      'SELECT * FROM todos WHERE json_extract(jsonData, \'\$.index\') = \'2\'',
      readsFrom: {todos}
    ).map((row){
      return TodoEntry.fromData(row.data, this);
    }).getSingle();
  }
```

I got this error
```
Error: no such column: jsonData
    at a.handleError (http://localhost:52411/sql-wasm.js:82:234)
    at a.prepare (http://localhost:52411/sql-wasm.js:80:361)
    at js.JsObject._fromJs.callMethod (http://localhost:52411/dart_sdk.js:60849:35)
    at sql_js.SqlJsDatabase.__.prepare (http://localhost:52411/packages/moor/src/web/sql_js.dart.lib.js:94:46)
    at moor_web._WebDelegate.new.runSelect (http://localhost:52411/packages/moor/moor_web.dart.lib.js:390:34)
    at http://localhost:52411/packages/moor/src/utils/lazy_database.dart.lib.js:2526:28
    at basic_lock.BasicLock.new.synchronized (http://localhost:52411/packages/synchronized/src/reentrant_lock.dart.lib.js:151:26)
    at synchronized.next (<anonymous>)
    at http://localhost:52411/dart_sdk.js:39191:33
    at _RootZone.runUnary (http://localhost:52411/dart_sdk.js:39048:58)
    at _FutureListener.thenAwait.handleValue (http://localhost:52411/dart_sdk.js:34034:29)
    at handleValueCallback (http://localhost:52411/dart_sdk.js:34594:49)
    at Function._propagateToListeners (http://localhost:52411/dart_sdk.js:34632:17)
    at _Future.new.[_complete] (http://localhost:52411/dart_sdk.js:34466:25)
    at _SyncCompleter.new.complete (http://localhost:52411/dart_sdk.js:33960:31)
    at _complete (http://localhost:52411/packages/synchronized/src/reentrant_lock.dart.lib.js:165:23)
    at basic_lock.BasicLock.new.synchronized (http://localhost:52411/packages/synchronized/src/reentrant_lock.dart.lib.js:173:13)
    at synchronized.next (<anonymous>)
    at http://localhost:52411/dart_sdk.js:39191:33
    at _RootZone.runUnary (http://localhost:52411/dart_sdk.js:39048:58)
    at _FutureListener.thenAwait.handleValue (http://localhost:52411/dart_sdk.js:34034:29)
    at handleValueCallback (http://localhost:52411/dart_sdk.js:34594:49)
    at Function._propagateToListeners (http://localhost:52411/dart_sdk.js:34632:17)
    at _Future.immediate.[_completeWithValue] (http://localhost:52411/dart_sdk.js:34474:23)
    at async._AsyncCallbackEntry.new.callback (http://localhost:52411/dart_sdk.js:34497:35)
    at Object._microtaskLoop (http://localhost:52411/dart_sdk.js:39335:13)
    at _startMicrotaskLoop (http://localhost:52411/dart_sdk.js:39341:13)
    at http://localhost:52411/dart_sdk.js:34848:9
```
Is there any better way to debug this? Thank you!

**Reply by thachnb85** [reply]
I already added this build.yaml to the same level of pubspec.yaml
```
targets:
  $default:
    builders:
      moor_generator:
        options:
          sqlite_modules:
            - json1
            - fts5
            - moor_ffi
```
And re-run:
```
flutter packages pub run build_runner build
```

I also re-tried the query with different format, but cannot find the table
```
  Future<TodoEntry> selectFirstItem() {
    return customSelect(
      'SELECT * FROM todos WHERE json_extract(todos.jsonData, "\$.index") = "2"',
      readsFrom: {todos}
    ).map((row){
      return TodoEntry.fromData(row.data, this);
    }).getSingle();
  }
```

Am I missing somethings?

**Reply by simolus3** [reply]
Moor converts columns declared in Dart to snake_case, so the column would probably be called `todos.json_data`.

For that reason I would probably use moor files for complex custom queries. Also note that (afaik) `json1` is not available by default in sqlite.js (since it looks like you're running on the web). You'd have to compile it yourself after editing the [Makefile](https://github.com/sql-js/sql.js/blob/master/Makefile) to enable json1.

**Reply by thachnb85** [reply]
Thank you for your reply, I have my own sql.js supports json1 extension, tested using javascript.

I have tried all of these queries 
```
        'SELECT * FROM todos WHERE json_extract(todos.json_data, \'\$.index\') = \'2\'',
        'SELECT * FROM todos WHERE json_extract(todos.json_data, "\$.index") = "2"',
```

But non of them work, throwing 
```
Error: malformed JSON
    at a.handleError (http://localhost:49387/sql-wasm.js:82:388)
    at a.step (http://localhost:49387/sql-wasm.js:73:409)
    at js.JsObject._fromJs.callMethod (http://localhost:49387/dart_sdk.js:60849:35)
    at sql_js.PreparedStatement.__.step (http://localhost:49387/packages/moor/src/web/sql_js.dart.lib.js:158:39)
    at moor_web._WebDelegate.new.runSelect (http://localhost:49387/packages/moor/moor_web.dart.lib.js:396:29)
    at http://localhost:49387/packages/moor/src/utils/lazy_database.dart.lib.js:2526:28
    at engines._BeforeOpeningExecutor.new._synchronized (http://localhost:49387/packages/moor/src/utils/lazy_database.dart.lib.js:2512:24)
    at _synchronized.next (<anonymous>)
    at runBody (http://localhost:49387/dart_sdk.js:39211:34)
    at Object._async [as async] (http://localhost:49387/dart_sdk.js:39242:7)
    at engines._BeforeOpeningExecutor.new.[_synchronized] (http://localhost:49387/packages/moor/src/utils/lazy_database.dart.lib.js:2508:20)
    at engines._BeforeOpeningExecutor.new.runSelect (http://localhost:49387/packages/moor/src/utils/lazy_database.dart.lib.js:2524:48)
    at runSelect.next (<anonymous>)
    at runBody (http://localhost:49387/dart_sdk.js:39211:34)
    at Object._async [as async] (http://localhost:49387/dart_sdk.js:39242:7)
    at engines._BeforeOpeningExecutor.new.runSelect (http://localhost:49387/packages/moor/src/utils/lazy_database.dart.lib.js:2522:20)
    at http://localhost:49387/packages/moor/src/utils/lazy_database.dart.lib.js:6409:104
    at http://localhost:49387/packages/moor/src/utils/lazy_database.dart.lib.js:1384:96
    at _rootRunUnary (http://localhost:49387/dart_sdk.js:39528:45)
    at async._CustomZone.new.runUnary (http://localhost:49387/dart_sdk.js:38697:14)
    at _FutureListener.then.handleValue (http://localhost:49387/dart_sdk.js:34034:29)
    at handleValueCallback (http://localhost:49387/dart_sdk.js:34594:49)
    at Function._propagateToListeners (http://localhost:49387/dart_sdk.js:34632:17)
    at _Future.immediate.[_completeWithValue] (http://localhost:49387/dart_sdk.js:34474:23)
    at http://localhost:49387/dart_sdk.js:34497:35
    at _rootRun (http://localhost:49387/dart_sdk.js:39520:14)
    at async._CustomZone.new.run (http://localhost:49387/dart_sdk.js:38690:14)
    at async._AsyncCallbackEntry.new.callback (http://localhost:49387/dart_sdk.js:38629:33)
    at Object._microtaskLoop (http://localhost:49387/dart_sdk.js:39335:13)
    at _startMicrotaskLoop (http://localhost:49387/dart_sdk.js:39341:13)
    at http://localhost:49387/dart_sdk.js:34848:9
```
The json is pretty simple as `{"index":"2"}`

![image](https://user-images.githubusercontent.com/21049491/110942092-ec23b800-82fe-11eb-8072-1330a8d2703b.png)

Do you have any idea about that query? TIA.

**Reply by Mike278** [reply]
The query with single quotes looks correct to me (a helpful mnemonic is **S**ingle quotes for **S**trings, **D**ouble quotes for **D**atabase names). 

"malformed JSON" indicates that `json_extract` is being called on invalid json. Could you check if the table has any rows with invalid json? The json in your screenshot looks correct, but even one invalid row will throw that error.

```sql
drop table if exists todos;
create table todos (title TEXT, json_data TEXT);
insert into todos values ('entry 1', '{"index":"1"}');
insert into todos values ('entry 2', '{"index":"2"}');
select * from todos where json_extract(todos.json_data, '$.index') = '2';
-- entry 2|{"index":"2"}
insert into todos values ('entry 3', '{"index":some invalid json}');
select * from todos where json_extract(todos.json_data, '$.index') = '3';
-- malformed JSON
```

Another way of using json like this that you might be interested in is to take advantage of sqlite's new support for [generated columns](https://sqlite.org/gencol.html) to automatically extract the json on insert. Something like:
```sql
select sqlite_version(); -- must be >= 3.31.0
drop table if exists todos;
create table todos (
  title TEXT,
  idx text as (json_extract(json_data, '$.index')),
  json_data TEXT
);
insert into todos (title, json_data) values ('entry 1', '{"index":"1"}');
insert into todos (title, json_data) values ('entry 2', '{"index":"2"}');
select * from todos where idx = '2';
-- entry 2|2|{"index":"2"}
```

**Reply by thachnb85** [reply]
Thanks @Mike278 Mike

I checked and printed every single row, the string for json is always correct, using double quotes:
```
{"index":"1"}
```
Starts with curly bracket, key and value are correct, and ending with curly bracket.
Using the sqlite command line, everything is correct, query is fine.
I just want to test the `moor` plugin if it really works with custom select query like above.

**Reply by thachnb85** [reply]
https://github.com/thachnb85/moor_shared
I have some example data and query, forked from moor_shared and added json test code so you can reproduce the issue above. Did moor modify the customQuery? @simolus3

**Reply by Mike278** [reply]
Ah I see - thanks for the repro! 

If you change `jsonData: Value('"{"index":"$i"}"')` to what you had earlier (`jsonData: Value('{"index":"$i"}')`), does it work?

**Reply by thachnb85** [reply]
I tried both, it doesn't work.

**Reply by thachnb85** [reply]
@Mike278  I found the issue!

I got the sql.js code from https://github.com/sql-js/sql.js
Then modify the Makefile to support JSON1, but after running the build command as they guided: `npm run rebuild`
When I use this output, JSON1 is supported.
```
sql-wasm-debug.js
sql-wasm-debug.wasm
```
When I use the release one
```
sql-wasm.js
sql-wasm.wasm
```
these files don't support JSON1

So nothing wrong with moor_shared project above.
The correct json text format should not begin with double quotes `jsonData: Value('"{"index":"$i"}"')`
Correct format of json is:`jsonData: Value('{"index":"$i"}')`
